### PR TITLE
Support PKCE flow

### PR DESCRIPTION
**Pull request by TheFacto**
👋 First time contributor.

I'm putting this out here as `Draft` as it's currently an idea, and I want to make sure it's something you'd like before I take the idea further and flesh it out with tests, etc.

A little background: https://developer.okta.com/blog/2019/08/22/okta-authjs-pkce#use-pkce-to-make-your-apps-more-secure

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -10,6 +10,9 @@ import {
   StrategyVerifyCallback,
 } from "remix-auth";
 import { v4 as uuid } from "uuid";
+import crypto from "crypto";
```
Comment (sergiodxa) [inline src/index.ts:13]: We can't use `crypto` because it doesn't work on Cloudflare

File: src/index.ts [hunk 2]
```diff
@@ -10,6 +10,9 @@ import {
...
+import randomstring from "randomstring";
+import base64url from "base64url";
```
Comment (sergiodxa) [inline src/index.ts:15]: Are these libs compatible with Cloudflare Worker runtime?

Comment (TheFacto) [reply src/index.ts:15]: Not sure, I'll do some digging. 🗒️

File: src/index.ts (hunk 1 above)
Comment (svengau) [reply src/index.ts:13]: Not 100% sure, but you could replace by `crypto-js`

File: src/index.ts (hunk 2 above)
Comment (svengau) [reply src/index.ts:15]: if you use `crypto-js`, you don't need this libs:

```
import cryptoJS from "crypto-js";
...
const codeVerifier = cryptoJS.lib.WordArray.random(96).toString(
  cryptoJS.enc.Base64url
);
const codeChallenge = cryptoJS
  .SHA256(code_verifier)
  .toString(cryptoJS.enc.Base64url);
```

File: package.json [hunk 3]
```diff
@@ -57,8 +58,10 @@
     "typescript": "^4.6.4"
   },
   "dependencies": {
+    "base64url": "^3.0.1",
     "debug": "^4.3.4",
-    "remix-auth": "^3.4.0",
+    "randomstring": "^1.2.3",
+    "remix-auth": "^3.2.2",
```
Comment (svengau) [inline package.json:64]: maybe remove this remix-auth version downgrade ?

**Comment by svengau** [comment]
Also, could you update the README file ? thx

**Comment by sergiodxa** [comment]
> why not make crypto a dependency injection?

While it will be easier for the package to do that, it means the dev using the strategy could provide functions that don't encrypt anything risking the security of their apps.

It's better if the strategy owns that so it can ensure it's safe, also it will be a breaking change for every user of the strategy and any of the substrategies.

**Comment by piotrkulpinski** [comment]
Hi guys, any progress on this? Is there any way to install the package from this specific branch?

**Comment by sergiodxa** [comment]
> Hi guys, any progress on this? Is there any way to install the package from this specific branch?

I've been working on a package to use OIDC with Web APIs, and I used it to create a [OIDCStrategy for Remix Auth](https://github.com/sergiodxa/web-oidc#use-with-remix-auth), while it doesn't support everything from OpenID Connect yet it does support PKCE.

**Comment by piotrkulpinski** [comment]
@sergiodxa That's great, thanks for doing that!

Unfortunately, in my case, this won't help as I'm trying to connect to Airtable which I believe doesn't support OIDC yet.

**Comment by sergiodxa** [comment]
The current strategy supports OAuth2 with PKCE too, OIDC works like OAuth2 with a few extra things, in your case instead of using the discovery feature of web-OIDC you can pass the issuer endpoints directly to the strategy.

**Comment by piotrkulpinski** [comment]
@sergiodxa Is there any example of using the package this way? I'm very new to the OIDC standard and couldn't find anything in the docs.

**Comment by sergiodxa** [comment]
@piotrkulpinski you can do this:

```ts
import { OIDCStrategy } from "web-oidc/remix";

authenticator.use(
  new OIDCStrategy(
    {
      issuer: {
        issuer: "https://auth.company.tld",
        authorization_endpoint: "https://auth.company.tld/authorize",
        token_endpoint: "https://auth.company.tld/oauth/token",
        userinfo_endpoint: "https://auth.company.tld/userinfo",
      },
      client_id: "CLIENT_ID",
      client_secret: "CLIENT_SECRET",
      redirect_uri: "https://www.company.tld/auth/callback",
      response_type: "code id_token",
    },
    async ({ profile, tokens, issuer, client }) => {
      return { profile, tokens };
    }
  )
);
```

You will need to replace the `issuer`, `authorization_endpoint`, `token_endpoint` and `userinfo_endpoint` with the values from Airtable. This is the minimum you need right now to use the strategy.

File: src/index.ts (hunk 2 above)
Comment (TheFacto) [reply src/index.ts:15]: @sergiodxa how would you feel about pulling in `crypto-js`?

File: package.json (hunk 3 above)
Comment (TheFacto) [reply package.json:64]: Good call, thank you.

**Comment by TheFacto** [comment]
@sergiodxa given your comments about using OIDC, do you feel we should pursue PKCE in `remix-auth-oauth2`? I'm still happy to pick this back up and utilize `crypto-js` to avoid the cloudfare conundrum you mentioned previously.

**Comment by sergiodxa** [comment]
@TheFacto if you want to work on it I will definitely merge it, PKCE doesn't work on all OAuth2 providers since it's a OIDC thing but it's ignored by the ones that doesn't support it so it doesn't break anything.

**Comment by Alexandre-Herve** [comment]
Hey there any update on this? I'm implementing a `remix-auth-lichess` strategy which requires PKCE. Maybe I can help with something?

**Comment by TheFacto** [comment]
@Alexandre-Herve I'm not sure the status here. I had pushed a few commits/tests and flagged as ready for review. I didn't want to pressure any maintainers to respect their time. Since then, I've picked up a few merge conflicts - I'm happy to get those resolved if there's still any interest to actually get this included.

**Comment by sergiodxa** [comment]
Support for this was implemented in #89

**Pull request closed without merging**